pallets/phala/src/stakepoolv2.rs

```diff
@@ -0,0 +1,3177 @@
+			<mining::pallet::Pallet<T>>::stop_mining(miner)?;
+			pool_info.cd_workers.push_back(worker.clone());
+			basepool::pallet::Pools::<T>::insert(pid, PoolProxy::StakePool(pool_info.clone()));
+			Ok(())
+		}
+		fn do_reclaim(
+			pid: u64,
+			sub_account: T::AccountId,
+			worker: WorkerPublicKey,
+			check_cooldown: bool,
+		) -> Result<(BalanceOf<T>, BalanceOf<T>), DispatchError> {
+			let (orig_stake, slashed) =
+				mining::Pallet::<T>::reclaim(sub_account.clone(), check_cooldown)?;
+			Self::handle_reclaim(pid, orig_stake, slashed);
+			Self::deposit_event(Event::<T>::WorkerReclaimed { pid, worker });
+			let mut pool_info = ensure_stake_pool::<T>(pid)?;
+			pool_info.remove_cd_worker(&worker);
+			basepool::pallet::Pools::<T>::insert(pid, PoolProxy::StakePool(pool_info.clone()));
+			Ok((orig_stake, slashed))
+		}
+
+		/// Adds up the newly received reward to `reward_acc`
+		fn handle_pool_new_reward(
+			pool_info: &mut StakePool<T::AccountId, BalanceOf<T>>,
+			rewards: BalanceOf<T>,
+		) {
+			if rewards > Zero::zero() {
+				mining::Pallet::<T>::withdraw_subsidy_pool(
+					&<T as pawnshop::Config>::PawnShopAccountId::get(),
+					rewards,
+				)
+				.expect("this should not happen");
+				if basepool::balance_close_to_zero(pool_info.basepool.total_shares) {
+					Self::deposit_event(Event::<T>::RewardDismissedNoShare {
+						pid: pool_info.basepool.pid,
+						amount: rewards,
+					});
+					return;
+				}
+				let commission = pool_info.payout_commission.unwrap_or_default() * rewards;
+
+				pallet_assets::Pallet::<T>::mint_into(
+					<T as pawnshop::Config>::PPhaAssetId::get(),
+					&pool_info.owner_account,
+					commission,
+				)
+				.expect("mint into should be success");
+				let to_distribute = rewards - commission;
+				pallet_assets::Pallet::<T>::mint_into(
+					<T as pawnshop::Config>::PPhaAssetId::get(),
+					&pool_info.basepool.pool_account_id,
+					to_distribute,
+				)
+				.expect("mint into should be success");
+				let distributed = if basepool::is_nondust_balance(to_distribute) {
+					pool_info.basepool.distribute_reward::<T>(to_distribute);
+					true
+				} else if to_distribute > Zero::zero() {
+					Self::deposit_event(Event::<T>::RewardDismissedDust {
+						pid: pool_info.basepool.pid,
+						amount: to_distribute,
+					});
+					false
+				} else {
+					false
+				};
+				if distributed || commission > Zero::zero() {
+					Self::deposit_event(Event::<T>::RewardReceived {
+						pid: pool_info.basepool.pid,
+						to_owner: commission,
+						to_stakers: to_distribute,
+					});
+				}
+			}
+		}
+
+		/// Called when worker was reclaimed.
+		///
+		/// After the cool down ends, worker was cleaned up, whose contributed balance would be
+		/// reset to zero.
+		fn handle_reclaim(pid: u64, orig_stake: BalanceOf<T>, slashed: BalanceOf<T>) {
+			let mut pool_info = ensure_stake_pool::<T>(pid).expect("Stake pool must exist; qed.");
+
+			let returned = orig_stake - slashed;
```

Comment:
Add a TODO to handle slash (move some locked token to the treasury)

Reply:
done

Reply:
Where?

Reply:
at lines just after this